### PR TITLE
first stab at floating-point comparators

### PR DESCRIPTION
**Pull request by desmonddak**
## Description & Motivation

This is a baseline for making arithmetic types have basic operations like comparison and negate.  This provides these operations 
for FloatingPoint.

## Related Issue(s)

#199

## Testing

All added operands have basic testing.

## Backwards-compatibility

> Is this a breaking change that will not be backwards-compatible? If yes, how so?

No

## Documentation

> Does the change require any updates to documentation? If so, where? Are they included?

No.  This is an extension to the API for these types but a very natural one whose documentation is in the class itself.

### CODE REVIEW

File: lib/src/arithmetic/signals/floating_point_logics/floating_point_logic.dart [hunk 1]
```diff
@@ -188,4 +188,91 @@ class FloatingPoint extends LogicStructure {
...
+      sign,
+      exponent,
+      mantissa,
+      input.explicitJBit,
+      name: input.name,
+    );
+  }
+
+  @override
+  FloatingPoint operator ~() => FloatingPoint.negate(this);
```
Comment (mkorbel1) [inline lib/src/arithmetic/signals/floating_point_logics/floating_point_logic.dart]: this does not seem correct -- this operator should be bitwise inversion right?

File: lib/src/arithmetic/signals/floating_point_logics/floating_point_logic.dart [hunk 2]
```diff
@@ -188,4 +188,91 @@ class FloatingPoint extends LogicStructure {
     final mantissa = Const(1, width: mantissaWidth);
     return FloatingPoint._(signLogic, exponent, mantissa, explicitJBit);
   }
+
+  /// Negate the [FloatingPoint].
+  factory FloatingPoint.negate(FloatingPoint input) {
```
Comment (mkorbel1) [inline lib/src/arithmetic/signals/floating_point_logics/floating_point_logic.dart]: why a factory instead of just a function or getter, or even lazy operation?  e.g.
```dart
late final FloatingPoint negated = ...
```

File: lib/src/arithmetic/signals/floating_point_logics/floating_point_logic.dart [hunk 3]
```diff
@@ -188,4 +188,91 @@ class FloatingPoint extends LogicStructure {
...
+      throw RohdHclException('Input must be floating point signal.');
+    }
+    if (explicitJBit || other.explicitJBit) {
+      throw RohdHclException(
+          'FloatingPoint with explicit J-bit cannot be compared.');
+    }
+    return mux(
+        isNaN | other.isNaN,
+        Const(0),
+        mux(this[-1], mux(other[-1], super.gt(other), Const(1)),
```
Comment (mkorbel1) [inline lib/src/arithmetic/signals/floating_point_logics/floating_point_logic.dart]: why use `[-1]` instead of `sign`?

File: lib/src/arithmetic/signals/floating_point_logics/floating_point_logic.dart [hunk 4]
```diff
@@ -188,4 +188,91 @@ class FloatingPoint extends LogicStructure {
...
+  @override
+  Logic lt(dynamic other) {
+    if (other is! FloatingPoint) {
+      throw RohdHclException('Input must be floating point signal.');
+    }
+    if (explicitJBit || other.explicitJBit) {
+      throw RohdHclException(
+          'FloatingPoint with explicit J-bit cannot be compared.');
+    }
+    return mux(
```
Comment (mkorbel1) [inline lib/src/arithmetic/signals/floating_point_logics/floating_point_logic.dart:242]: what if the two `FloatingPoint`s have different exponent/mantissa widths?

File: test/arithmetic/floating_point/floating_point_test.dart [hunk 5]
```diff
@@ -110,4 +112,72 @@ void main() {
...
+        expect(
+            fp1.neq(fp2).value.toBool(), isTrue); // This will use Logic.neq()
+      } else if (val1.toDouble() > val2.toDouble()) {
+        expect(fp2.lt(fp1).value.toBool(), isTrue);
+        expect(fp1.gte(fp2).value.toBool(), isTrue);
+        expect(
+            fp1.neq(fp2).value.toBool(), isTrue); // This will use Logic.neq()
+      } else {
+        // rare that the two numbers would collide but just to be safe
+        expect(fp1.eq(fp2).value.toBool(), isTrue);
```
Comment (mkorbel1) [inline test/arithmetic/floating_point/floating_point_test.dart]: TODO: rest of the testing for the operators?  we should definitely test equality right?

Comment (desmonddak) [reply test/arithmetic/floating_point/floating_point_test.dart]: So LogicStructure does not support operator != or operator ==.
When I try to add, it complains in FloatingPointValue about need for non-nullable type and a hashcode.

operator != is not overridable apparently.

operator -()   -- the linter cannot decide if this should have override or not!

Comment (mkorbel1) [reply test/arithmetic/floating_point/floating_point_test.dart]: I never played with the `-` operator, interesting.

You should *not* override the `==` operator since we want that to stay as object equality.  I meant we should be sure to test that `eq` works, and also some specific corner cases (e.g. NaN, infinities, etc.) and some of the exceptions that we're checking (mismatch widths, jbit-ness, etc.)

File: lib/src/arithmetic/signals/floating_point_logics/floating_point_logic.dart [hunk 6]
```diff
@@ -188,4 +188,72 @@ class FloatingPoint extends LogicStructure {
...
+      );
+
+  /// Negate the [FloatingPoint].
+  FloatingPoint operator -() => negate();
+
+  @override
+  Logic operator >(dynamic other) => gt(other);
+  @override
+  Logic operator >=(dynamic other) => gte(other);
+  @override
```
Comment (mkorbel1) [inline lib/src/arithmetic/signals/floating_point_logics/floating_point_logic.dart]: what is this floating `@override` doing here? :)

File: test/arithmetic/floating_point/floating_point_test.dart [hunk 7]
```diff
@@ -110,4 +112,131 @@ void main() {
...
+        expect(fp1.neq(fp2).value.toBool(), isFalse);
+      }
+    }
+  });
+
+  // TODO(desmonddak): convert these tests to methods
+
+  BigInt toBigInt(FloatingPointValue fpv) {
```
Comment (mkorbel1) [inline test/arithmetic/floating_point/floating_point_test.dart]: what are these to/from bigint doing in here?

Comment (desmonddak) [reply test/arithmetic/floating_point/floating_point_test.dart]: Trial code.  I discovered I can simply use converters to and from FixedPointValue and cover issues #112 , #113.

I didn't anticipate needing to make changes to both Fixed and Float at the same time, so this was experimental code and I'll delete.

File: lib/src/arithmetic/signals/floating_point_logics/floating_point_logic.dart (hunk 6 above)
Comment (desmonddak) [reply lib/src/arithmetic/signals/floating_point_logics/floating_point_logic.dart]: Ugh.

File: test/arithmetic/floating_point/floating_point_test.dart (hunk 5 above)
Comment (desmonddak) [reply test/arithmetic/floating_point/floating_point_test.dart]: I'll be adding the corner case testing.  Infinities are always equal (modulo the sign), but NaNs are never equal.

File: test/arithmetic/floating_point/floating_point_test.dart (hunk 7 above)
Comment (desmonddak) [reply test/arithmetic/floating_point/floating_point_test.dart]: The direction I am going:  conversions between FP and FX, with rounding logic will allow us to validate very wide FP operations.

File: test/arithmetic/floating_point/floating_point_test.dart (hunk 5 above)
Comment (desmonddak) [reply test/arithmetic/floating_point/floating_point_test.dart]: Should we not do this (already in the code):

floating_point_value.dart
```dart
  @override
  bool operator ==(Object other) {
    if (other is! FloatingPointValue) {
      return false;
    }
    if ((exponent.width != other.exponent.width) |
        (mantissa.width != other.mantissa.width)) {
      return false;
    }
    if (isNaN != other.isNaN) {
      return false;
    }
    if (isAnInfinity != other.isAnInfinity) {
      return false;
    }
    if (isAnInfinity) {
      return sign == other.sign;

Comment (desmonddak) [reply test/arithmetic/floating_point/floating_point_test.dart]: Another weird case:  I thought we could not override <= or <:

fixed_point_value.dart:
```dart
  @override
  bool operator ==(Object other) {
    if (other is! FloatingPointValue) {
      return false;
    }
    if ((exponent.width != other.exponent.width) |
        (mantissa.width != other.mantissa.width)) {
      return false;
    }
    if (isNaN != other.isNaN) {
      return false;
    }
    if (isAnInfinity != other.isAnInfinity) {
      return false;
    }
    if (isAnInfinity) {
      return sign == other.sign;

Comment (mkorbel1) [reply test/arithmetic/floating_point/floating_point_test.dart]: yes we should on the `*Value`s, but not on the things implementing `Logic`.  The values are sort of like immutable software primitives, whereas signals are objects that change value over time during simulation.

File: lib/src/arithmetic/signals/floating_point_logics/floating_point_logic.dart [hunk 8]
```diff
@@ -78,7 +78,7 @@ class FloatingPoint extends LogicStructure {
   /// Return true if the J-bit is explicitly represented in the mantissa.
   bool get explicitJBit => _explicitJBit;
 
-  late final bool _explicitJBit;
+  final bool _explicitJBit;
```
Comment (mkorbel1) [inline lib/src/arithmetic/signals/floating_point_logics/floating_point_logic.dart:81]: since it's final, we could actually just skip the private version and getter, just have `final bool explicitJBit`

File: lib/src/arithmetic/signals/floating_point_logics/floating_point_logic.dart [hunk 9]
```diff
@@ -190,72 +190,63 @@ class FloatingPoint extends LogicStructure {
...
+        Logic()..gets(~sign),
+        Logic(width: exponent.width)..gets(exponent),
+        Logic(width: mantissa.width)..gets(mantissa),
+        explicitJBit,
+        name: name,
+      );
 
+  /// Negate the [FloatingPoint].
+  FloatingPoint operator -() => negate();
+  // ignore the lint warning about overriding the '-' operator.
```
Comment (mkorbel1) [inline lib/src/arithmetic/signals/floating_point_logics/floating_point_logic.dart:203]: what lint warning?  does the analyzer flag a lint or not, because it should be failing if you don't have the right ignore syntax

File: lib/src/arithmetic/signals/floating_point_logics/floating_point_logic.dart [hunk 10]
```diff
@@ -188,4 +188,77 @@ class FloatingPoint extends LogicStructure {
     final mantissa = Const(1, width: mantissaWidth);
     return FloatingPoint._(signLogic, exponent, mantissa, explicitJBit);
   }
+
+  /// Negate the [FloatingPoint].
+  FloatingPoint negate() => FloatingPoint._(
```
Comment (mkorbel1) [inline lib/src/arithmetic/signals/floating_point_logics/floating_point_logic.dart:193]: is there anything special to consider for NaN/infinity here?

File: lib/src/arithmetic/values/floating_point_values/floating_point_value.dart [hunk 11]
```diff
@@ -492,19 +492,45 @@ class FloatingPointValue implements Comparable<FloatingPointValue> {
...
   }
 
   /// Negate operation for [FloatingPointValue].
-  FloatingPointValue negate() => clonePopulator().populate(
+  FloatingPointValue negated() => clonePopulator().populate(
       sign: sign.isZero ? LogicValue.one : LogicValue.zero,
       exponent: exponent,
       mantissa: mantissa);
 
+  // Move this to FixedPointValuePopulator.ofFloatingPointValue.
```
Comment (mkorbel1) [inline lib/src/arithmetic/values/floating_point_values/floating_point_value.dart]: resolve this comment?

also, wouldn't we always want a fixed width `FixedPointValue` per widths of the `FloatingPointValue` components rather than the width changing based on values?

File: lib/src/arithmetic/values/floating_point_values/floating_point_value_populator.dart [hunk 12]
```diff
@@ -205,25 +205,56 @@ class FloatingPointValuePopulator<FpvType extends FloatingPointValue> {
...
           sign: LogicValue.ofBigInt(sign ? BigInt.one : BigInt.zero, 1),
           exponent: LogicValue.ofBigInt(exponent, exponentWidth),
           mantissa: LogicValue.ofBigInt(mantissa, mantissaWidth));
 
-  /// [FloatingPointValue] constructor from a set of [int]s of the binary
-  /// representation and the size of the exponent and mantissa
+  /// [FloatingPointValue] populator from a [FixedPointValue],
+  FpvType ofFixedPointValue(FixedPointValue fxv,
+      {FloatingPointRoundingMode mode =
+          FloatingPointRoundingMode.roundNearestEven}) {
```
Comment (mkorbel1) [inline lib/src/arithmetic/values/floating_point_values/floating_point_value_populator.dart]: do you want to add a check that the rounding mode is not changed until we support more?

File: lib/src/utils.dart [hunk 13]
```diff
@@ -10,6 +10,19 @@ import 'package:rohd/rohd.dart';
 /// Computes the bit width needed to store [w] addresses.
 int log2Ceil(int w) => (log(w) / log(2)).ceil();
 
+/// Finds the index of the first set bit in a [BigInt] number.
+/// Returns length + 1 if there are no set bits (i.e., the number is zero).
```
Comment (mkorbel1) [inline lib/src/utils.dart:14]: isn't this an infinite loop if `inNumber` is 0?

wouldn't `null` be a better return value if none is found?

also, is this a utility we want to expose in the public API?  we have other things like `Find` which can also find the `n`th instance of a 1 *or* 0, plus an error signal on the hardware side.  Maybe a more complete extension to `LogicValue` would be more valuable as a public API?

File: lib/src/arithmetic/values/floating_point_values/floating_point_value.dart [hunk 14]
```diff
@@ -492,19 +492,45 @@ class FloatingPointValue implements Comparable<FloatingPointValue> {
...
+    var fxdMantissa = [
+      if (isNormal()) LogicValue.one else LogicValue.zero,
+      mantissa
+    ].swizzle().zeroExtend(bias * 2 + mantissaWidth + 2);
+    fxdMantissa = sign == LogicValue.one
+        ? ~fxdMantissa + 1
+        : fxdMantissa; // make sure the sign is correct
+    final shift = exponent.toInt() - 1;
+    final shiftedFxdMantissa =
+        exponent.toInt() > 0 ? fxdMantissa << shift : fxdMantissa;
```
Comment (sweeksBigBlue) [inline lib/src/arithmetic/values/floating_point_values/floating_point_value.dart:519]: Are we shifting out the unbiased exponent here? I would think you would need to debias the exponent but maybe I'm missing something further here?

File: lib/src/arithmetic/signals/floating_point_logics/floating_point_logic.dart (hunk 9 above)
Comment (desmonddak) [reply lib/src/arithmetic/signals/floating_point_logics/floating_point_logic.dart:203]: The analyzer flags a lint error, but it doesn't fail in CI.  It is a false error, the operator works.  I don't have to have this operator.  I see a blue underline in my VSCode with the message that this operator needs an `@override` annotation
image

When I add override, it still complains, probably because LogicValue doesn't have a '-' operator:
image

File: lib/src/arithmetic/signals/floating_point_logics/floating_point_logic.dart (hunk 10 above)
Comment (desmonddak) [reply lib/src/arithmetic/signals/floating_point_logics/floating_point_logic.dart:193]: Oddly, no.   The standard says nothing about having a signed NaN.  Negating infinity makes sense as does negating zero.  So I could not think of anything to check here.

File: lib/src/arithmetic/values/floating_point_values/floating_point_value.dart (hunk 11 above)
Comment (desmonddak) [reply lib/src/arithmetic/values/floating_point_values/floating_point_value.dart]: The width of FixedPointValue would be fixed but very wide to make FP conversion lossless.  But that width is a function of the FP width  (bias * 2 + mantissaWidth + 2):   the full range of sliding an FP number up and down the scale of exponents plus its mantissa plus a jbit plus a sign bit.

So I thought it made sense to do a .toFixedPointValue rather than doing an .ofFloatingPointValue since we need to compute the widths dependent on the FP exponent/mantissa widths (not values).

Comment (desmonddak) [reply lib/src/arithmetic/values/floating_point_values/floating_point_value.dart]: Maybe I didn't understand this question, but this function is dependent on FP component widths and is written to handle that -- did you see something value-based in the method?

Comment (desmonddak) [reply lib/src/arithmetic/values/floating_point_values/floating_point_value.dart]: Ah, now I see.  My comment was misleading and confused.  I do NOT want to build anything that is value dependent.  I was considering it when I could not figure out just how wide I had to go to cover the full range.

File: lib/src/utils.dart (hunk 13 above)
Comment (desmonddak) [reply lib/src/utils.dart:14]: I like the idea of a LogicValue extension.  We could mirror 'Find' in the Value side.  Let me take a look.

File: lib/src/arithmetic/values/floating_point_values/floating_point_value.dart (hunk 14 above)
Comment (desmonddak) [reply lib/src/arithmetic/values/floating_point_values/floating_point_value.dart:519]: No, its really weird:  I am using the unbiased exponent of 00...001 to be fully right-shifted mantissa and then 11....10 to be the full left-shifted mantissa stored in a really wide FixedPoint integer.  We would set n/m to be around the midpoint of this super-wide LogicStructure.

File: lib/src/utils.dart (hunk 13 above)
Comment (desmonddak) [reply lib/src/utils.dart:14]: For now I added code to match my other leading1 detectors.  For some reason I have not tripped on this bug.
It makes sense to move to Find.

File: lib/src/arithmetic/values/floating_point_values/floating_point_value.dart (hunk 14 above)
Comment (sweeksBigBlue) [reply lib/src/arithmetic/values/floating_point_values/floating_point_value.dart:519]: Oh okay, I would for sure need that explained to me 😆

**Comment by desmonddak** [comment]
I converted this PR to a draft as there is a lot to fix.

File: lib/src/arithmetic/signals/floating_point_logics/floating_point_logic.dart (hunk 9 above)
Comment (mkorbel1) [reply lib/src/arithmetic/signals/floating_point_logics/floating_point_logic.dart:203]: hmm weird!  i wonder if it's an analyzer bug?

File: lib/src/arithmetic/values/floating_point_values/floating_point_value.dart (hunk 14 above)
Comment (desmonddak) [reply lib/src/arithmetic/values/floating_point_values/floating_point_value.dart:519]: I answered poorly.  yes, we are shifting the mantissa by the unbiased exponent.  That way, the tiniest numbers are all the way to the right (from mantissa-1 to 0).  Then the largest numbers are shifted as far left as we can so the mantissa sits at the left edge down to left edge - mantissa-width.  that makes the fixed-point around 2*bias + mantissa-width wide.
